[Typescript] Incompatible with non-synthetic default imports

When using chakra-ui in a typescript project, it won't build without `allowSyntheticDeafultImports`.

With v0.8, an import of `styled-system` fails. With v1 (rc2), imports of react and others fail.

eg.

```
node_modules/@chakra-ui/core/dist/Popover/index.d.ts:3:8 - error TS1192: Module '"/home/joel/dev/ps-evo/node_modules/@types/styled-system/index"' has no default export.

3 import StyledSystem from "styled-system";
```

**Steps to Reproduce** Clear and concise reproduction instructions are important
for us to be able to triage your issue in a timely manner.

1. Make a normal typescript project, using default options (not presets like CRA or others).
2. Build with chakra-ui as a dependency.

**Suggested solution(s)**
I'm generally in favor of non-synthetic imports (`import * as React`), but I'm not sure of alternative ways to support this. I don't want to force consumers of chakra-ui to use interopt flags or other things that end up increasing bundle sizes or making things non-esm compliant.

I'd be just as happy though, if there's a simpler way to support consumers with either configuration.

--- a/examples/nextjs-typescript/tsconfig.json
+++ b/examples/nextjs-typescript/tsconfig.json
@@ -16,7 +16,8 @@
     "resolveJsonModule": true,
     "skipLibCheck": true,
     "strict": true,
-    "target": "esnext"
+    "target": "esnext",
+    "allowSyntheticDefaultImports": true
   },
   "exclude": ["node_modules"],
   "include": ["**/*.ts", "**/*.tsx"]

--- a/packages/accordion/src/accordion.tsx
+++ b/packages/accordion/src/accordion.tsx
@@ -19,7 +19,7 @@ import {
   __DEV__,
   Dict,
 } from "@chakra-ui/utils"
-import React, { useMemo } from "react"
+import * as React from "react"
 import {
   AccordionProvider,
   useAccordion,
@@ -58,7 +58,7 @@ export const Accordion = forwardRef<AccordionProps, "div">(function Accordion(
 
   const { children, htmlProps, ...context } = useAccordion(_props)
 
-  const _context = useMemo(
+  const _context = React.useMemo(
     () => ({ ...context, reduceMotion: !!props.reduceMotion }),
     [context, props.reduceMotion],
   )
@@ -113,7 +113,7 @@ export const AccordionItem = forwardRef<AccordionItemProps, "div">(
     const { htmlProps, ...context } = useAccordionItem(props)
 
     const styles = useStyles()
-    const _context = useMemo(() => context, [context])
+    const _context = React.useMemo(() => context, [context])
 
     return (
       <AccordionItemProvider value={_context}>

--- a/packages/alert/src/alert.tsx
+++ b/packages/alert/src/alert.tsx
@@ -8,7 +8,7 @@ import {
   useStyles,
 } from "@chakra-ui/system"
 import { createContext, cx } from "@chakra-ui/utils"
-import React from "react"
+import * as React from "react"
 import { CheckIcon, InfoIcon, WarningIcon } from "./icons"
 
 const STATUSES = {

--- a/packages/avatar/src/avatar-group.tsx
+++ b/packages/avatar/src/avatar-group.tsx
@@ -8,7 +8,7 @@ import {
   forwardRef,
 } from "@chakra-ui/system"
 import { getValidChildren, __DEV__, cx } from "@chakra-ui/utils"
-import React, { ReactNode, cloneElement } from "react"
+import * as React from "react"
 import { baseStyle } from "./avatar"
 
 interface AvatarGroupOptions {
@@ -17,7 +17,7 @@ interface AvatarGroupOptions {
    *
    * Ideally should be `Avatar` and `MoreIndicator` components
    */
-  children: ReactNode
+  children: React.ReactNode
   /**
    * The space between the avatars in the group.
    */
@@ -69,7 +69,7 @@ export const AvatarGroup = forwardRef<AvatarGroupProps, "div">(
     const clones = reversedChildren.map((child, index) => {
       const isFirstAvatar = index === 0
 
-      return cloneElement(child, {
+      return React.cloneElement(child, {
         mr: isFirstAvatar ? 0 : spacing,
         size: props.size,
         borderColor: child.props.borderColor || borderColor,

--- a/packages/avatar/src/avatar.tsx
+++ b/packages/avatar/src/avatar.tsx
@@ -12,7 +12,7 @@ import {
   forwardRef,
 } from "@chakra-ui/system"
 import { cx, __DEV__ } from "@chakra-ui/utils"
-import React, { cloneElement, ReactElement, ReactNode } from "react"
+import * as React from "react"
 
 interface AvatarOptions {
   /**
@@ -35,7 +35,7 @@ interface AvatarOptions {
   /**
    * The badge at the bottom right corner of the avatar.
    */
-  children?: ReactNode
+  children?: React.ReactNode
   /**
    * The image url of the `Avatar`
    */
@@ -56,7 +56,7 @@ interface AvatarOptions {
    * The default avatar used as fallback when `name`, and `src`
    * is not specified.
    */
-  icon?: ReactElement
+  icon?: React.ReactElement
   /**
    * Function to get the initials to display
    */
@@ -254,7 +254,7 @@ const AvatarImage: React.FC<AvatarImageProps> = ({
         name={name}
       />
     ) : (
-      cloneElement(icon, { role: "img" })
+      React.cloneElement(icon, { role: "img" })
     )
   }
 

--- a/packages/breadcrumb/src/breadcrumb.tsx
+++ b/packages/breadcrumb/src/breadcrumb.tsx
@@ -10,7 +10,7 @@ import {
   useStyles,
 } from "@chakra-ui/system"
 import { cx, getValidChildren, __DEV__ } from "@chakra-ui/utils"
-import React, { cloneElement } from "react"
+import * as React from "react"
 
 export interface BreadcrumbSeparatorProps extends PropsOf<typeof chakra.div> {
   spacing?: SystemProps["mx"]
@@ -109,13 +109,13 @@ export const BreadcrumbItem = forwardRef<BreadcrumbItemProps, "li">(
 
     const clones = validChildren.map((child) => {
       if (child.type === BreadcrumbLink) {
-        return cloneElement(child, {
+        return React.cloneElement(child, {
           isCurrentPage,
         })
       }
 
       if (child.type === BreadcrumbSeparator) {
-        return cloneElement(child, {
+        return React.cloneElement(child, {
           spacing,
           children: child.props.children || separator,
         })
@@ -188,7 +188,7 @@ export const Breadcrumb = forwardRef<BreadcrumbProps, "nav">(
     const count = validChildren.length
 
     const clones = validChildren.map((child, index) =>
-      cloneElement(child, {
+      React.cloneElement(child, {
         separator,
         spacing,
         isLastChild: count === index + 1,

--- a/packages/button/src/button-group.tsx
+++ b/packages/button/src/button-group.tsx
@@ -7,7 +7,7 @@ import {
   PropsOf,
 } from "@chakra-ui/system"
 import { createContext, cx, __DEV__ } from "@chakra-ui/utils"
-import React, { useMemo } from "react"
+import * as React from "react"
 
 export interface ButtonGroupProps
   extends PropsOf<typeof chakra.div>,
@@ -56,7 +56,7 @@ export const ButtonGroup = forwardRef<ButtonGroupProps, "div">(
 
     const _className = cx("chakra-button__group", className)
 
-    const context = useMemo(
+    const context = React.useMemo(
       () => ({ size, colorScheme, variant, isDisabled }),
       [size, colorScheme, variant, isDisabled],
     )

--- a/packages/button/src/button.tsx
+++ b/packages/button/src/button.tsx
@@ -9,7 +9,7 @@ import {
   useStyleConfig,
 } from "@chakra-ui/system"
 import { cx, dataAttr, merge, __DEV__ } from "@chakra-ui/utils"
-import React, { ReactElement, isValidElement, cloneElement } from "react"
+import * as React from "react"
 import { useButtonGroup } from "./button-group"
 
 export interface ButtonOptions {
@@ -41,11 +41,11 @@ export interface ButtonOptions {
   /**
    * If added, the button will show an icon before the button's label.
    */
-  leftIcon?: ReactElement
+  leftIcon?: React.ReactElement
   /**
    * If added, the button will show an icon after the button's label.
    */
-  rightIcon?: ReactElement
+  rightIcon?: React.ReactElement
   /**
    * The space between the button icon and label.
    */
@@ -53,7 +53,7 @@ export interface ButtonOptions {
   /**
    * Replace the spinner component when `isLoading` is set to `true`
    */
-  spinner?: ReactElement
+  spinner?: React.ReactElement
 }
 
 export interface ButtonProps
@@ -149,8 +149,8 @@ if (__DEV__) {
 const ButtonIcon: React.FC<PropsOf<typeof chakra.span>> = (props) => {
   const { children, className, ...rest } = props
 
-  const _children = isValidElement(children)
-    ? cloneElement(children, {
+  const _children = React.isValidElement(children)
+    ? React.cloneElement(children, {
         "aria-hidden": true,
         focusable: false,
       })

--- a/packages/button/src/icon-button.tsx
+++ b/packages/button/src/icon-button.tsx
@@ -1,5 +1,5 @@
 import { __DEV__ } from "@chakra-ui/utils"
-import React, { ReactElement, isValidElement, cloneElement } from "react"
+import * as React from "react"
 import { Button, ButtonProps } from "./button"
 import { forwardRef } from "@chakra-ui/system"
 
@@ -16,7 +16,7 @@ export interface IconButtonProps extends BaseButtonProps {
   /**
    * The icon to be used in the button.
    */
-  icon?: ReactElement
+  icon?: React.ReactElement
   /**
    * If `true`, the button will be perfectly round. Else, it'll be slightly round
    */
@@ -35,8 +35,8 @@ export const IconButton = forwardRef<IconButtonProps, "button">(
      * Passing the icon as prop or children should work
      */
     const element = icon || children
-    const _children = isValidElement(element)
-      ? cloneElement(element as any, {
+    const _children = React.isValidElement(element)
+      ? React.cloneElement(element as any, {
           "aria-hidden": true,
           focusable: false,
         })

--- a/packages/checkbox/src/checkbox-group.tsx
+++ b/packages/checkbox/src/checkbox-group.tsx
@@ -1,6 +1,6 @@
 import { ThemingProps } from "@chakra-ui/system"
 import { createContext, __DEV__ } from "@chakra-ui/utils"
-import React, { useMemo, ReactNode } from "react"
+import * as React from "react"
 import {
   useCheckboxGroup,
   UseCheckboxGroupProps,
@@ -10,7 +10,7 @@ import {
 export interface CheckboxGroupProps
   extends UseCheckboxGroupProps,
     Omit<ThemingProps, "orientation"> {
-  children?: ReactNode
+  children?: React.ReactNode
 }
 
 export interface CheckboxGroupContext
@@ -36,7 +36,7 @@ export const CheckboxGroup: React.FC<CheckboxGroupProps> = (props) => {
   const { colorScheme, size, variant, children } = props
   const { value, onChange } = useCheckboxGroup(props)
 
-  const group = useMemo(
+  const group = React.useMemo(
     () => ({
       size,
       onChange,

--- a/packages/icon/src/create-icon.tsx
+++ b/packages/icon/src/create-icon.tsx
@@ -1,4 +1,4 @@
-import React from "react"
+import * as React from "react"
 import { Icon, IconProps } from "./icon"
 import { __DEV__ } from "@chakra-ui/utils"
 import { forwardRef, PropsOf } from "@chakra-ui/system"

--- a/packages/input/src/input-group.tsx
+++ b/packages/input/src/input-group.tsx
@@ -8,7 +8,7 @@ import {
   forwardRef,
 } from "@chakra-ui/system"
 import { cx, __DEV__, getValidChildren } from "@chakra-ui/utils"
-import React, { cloneElement, useRef } from "react"
+import * as React from "react"
 
 export interface InputGroupProps
   extends PropsOf<typeof chakra.div>,
@@ -20,7 +20,7 @@ export const InputGroup = forwardRef<InputGroupProps, "div">(
     const { children, className, ...rest } = omitThemingProps(props)
 
     const _className = cx("chakra-input__group", className)
-    const stylesRef = useRef<InputGroupProps>({})
+    const stylesRef = React.useRef<InputGroupProps>({})
 
     const validChildren = getValidChildren(children)
 
@@ -51,8 +51,8 @@ export const InputGroup = forwardRef<InputGroupProps, "div">(
       const { pl, paddingLeft, pr, paddingRight } = child.props
 
       return child.type.id !== "Input"
-        ? cloneElement(child, theming)
-        : cloneElement(child, {
+        ? React.cloneElement(child, theming)
+        : React.cloneElement(child, {
             ...theming,
             paddingLeft: pl ?? paddingLeft ?? stylesRef.current?.paddingLeft,
             paddingRight: pr ?? paddingRight ?? stylesRef.current?.paddingRight,

--- a/packages/layout/src/aspect-ratio.tsx
+++ b/packages/layout/src/aspect-ratio.tsx
@@ -1,5 +1,5 @@
 import { cx, __DEV__, mapResponsive } from "@chakra-ui/utils"
-import React, { Children } from "react"
+import * as React from "react"
 import { chakra, PropsOf, ResponsiveValue, forwardRef } from "@chakra-ui/system"
 
 interface AspectRatioOptions {
@@ -26,7 +26,7 @@ export const AspectRatio = forwardRef<AspectRatioProps, "div">(
     const { ratio = 4 / 3, children, className, ...rest } = props
 
     // enforce single child
-    const child = Children.only(children)
+    const child = React.Children.only(children)
 
     const _className = cx("chakra-aspect-ratio", className)
 

--- a/packages/menu/src/menu.tsx
+++ b/packages/menu/src/menu.tsx
@@ -16,7 +16,7 @@ import {
   runIfFn,
   __DEV__,
 } from "@chakra-ui/utils"
-import React, { ReactElement, useMemo } from "react"
+import * as React from "react"
 import {
   MenuProvider,
   useMenu,
@@ -45,7 +45,7 @@ export const Menu: React.FC<MenuProps> = (props) => {
   const realProps = omitThemingProps(props)
 
   const ctx = useMenu(realProps)
-  const context = useMemo(() => ctx, [ctx])
+  const context = React.useMemo(() => ctx, [ctx])
 
   return (
     <MenuProvider value={context}>
@@ -182,7 +182,7 @@ interface MenuItemOptions
   /**
    * The icon to render before the menu item's label.
    */
-  icon?: ReactElement
+  icon?: React.ReactElement
   /**
    * The spacing between the icon and menu item's label
    */
@@ -240,7 +240,7 @@ if (__DEV__) {
 export interface MenuItemOptionProps
   extends UseMenuOptionOptions,
     Omit<PropsOf<typeof StyledMenuItem>, keyof UseMenuOptionOptions> {
-  icon?: ReactElement
+  icon?: React.ReactElement
   iconSpacing?: SystemProps["mr"]
 }
 

--- a/packages/modal/src/modal.tsx
+++ b/packages/modal/src/modal.tsx
@@ -11,13 +11,13 @@ import {
   useStyles,
 } from "@chakra-ui/system"
 import { callAllHandlers, cx, __DEV__ } from "@chakra-ui/utils"
-import React, { ReactNode, useEffect } from "react"
+import * as React from "react"
 import { RemoveScroll } from "react-remove-scroll"
 import { ModalContextProvider, useModalContext } from "./context"
 import { useModal, UseModalProps } from "./use-modal"
 
 export interface ModalProps extends UseModalProps, ThemingProps {
-  children?: ReactNode
+  children?: React.ReactNode
   /**
    *  If `true`, the modal will be centered on screen.
    * @default false
@@ -206,7 +206,7 @@ export const ModalHeader = forwardRef<ModalHeaderProps, "header">(
      * Notify us if this component was rendered or used
      * so we can append `aria-labelledby` automatically
      */
-    useEffect(() => {
+    React.useEffect(() => {
       setHeaderMounted(true)
       return () => setHeaderMounted(false)
     }, [setHeaderMounted])
@@ -253,7 +253,7 @@ export const ModalBody = forwardRef<ModalBodyProps, "div">(function ModalBody(
    * Notify us if this component was rendered or used
    * so we can append `aria-describedby` automatically
    */
-  useEffect(() => {
+  React.useEffect(() => {
     setBodyMounted(true)
     return () => setBodyMounted(false)
   }, [setBodyMounted])

--- a/packages/popover/src/popover.tsx
+++ b/packages/popover/src/popover.tsx
@@ -16,7 +16,7 @@ import {
   runIfFn,
   __DEV__,
 } from "@chakra-ui/utils"
-import React, { Children, cloneElement, useEffect } from "react"
+import * as React from "react"
 import { usePopover, UsePopoverProps, UsePopoverReturn } from "./use-popover"
 
 const [PopoverProvider, usePopoverContext] = createContext<UsePopoverReturn>({
@@ -70,9 +70,9 @@ if (__DEV__) {
  */
 export const PopoverTrigger: React.FC = (props) => {
   // enforce a single child
-  const child: any = Children.only(props.children)
+  const child: any = React.Children.only(props.children)
   const { getTriggerProps } = usePopoverContext()
-  return cloneElement(child, getTriggerProps(child.props, child.ref))
+  return React.cloneElement(child, getTriggerProps(child.props, child.ref))
 }
 
 if (__DEV__) {
@@ -122,7 +122,7 @@ export const PopoverHeader = forwardRef<PopoverHeaderProps, "header">(
   function PopoverHeader(props, ref) {
     const { headerId, setHasHeader } = usePopoverContext()
 
-    useEffect(() => {
+    React.useEffect(() => {
       setHasHeader.on()
       return () => setHasHeader.off()
     }, [setHasHeader])
@@ -155,7 +155,7 @@ export const PopoverBody = forwardRef<PopoverBodyProps, "div">(
   function PopoverBody(props, ref) {
     const { bodyId, setHasBody } = usePopoverContext()
 
-    useEffect(() => {
+    React.useEffect(() => {
       setHasBody.on()
       return () => setHasBody.off()
     }, [setHasBody])

--- a/packages/tabs/src/tabs.tsx
+++ b/packages/tabs/src/tabs.tsx
@@ -9,7 +9,7 @@ import {
   useStyles,
 } from "@chakra-ui/system"
 import { cx, __DEV__ } from "@chakra-ui/utils"
-import React, { ReactNode, useMemo } from "react"
+import * as React from "react"
 import {
   TabsProvider,
   useTab,
@@ -39,7 +39,7 @@ export interface TabsProps
     ThemingProps,
     Omit<PropsOf<typeof chakra.div>, "onChange">,
     TabsOptions {
-  children: ReactNode
+  children: React.ReactNode
 }
 
 /**
@@ -53,7 +53,7 @@ export const Tabs = forwardRef<TabsProps, "div">(function Tabs(props, ref) {
   const { children, className, ...otherProps } = omitThemingProps(props)
 
   const { htmlProps, ...ctx } = useTabs(otherProps)
-  const context = useMemo(() => ctx, [ctx])
+  const context = React.useMemo(() => ctx, [ctx])
 
   return (
     <TabsProvider value={context}>

